lib/plugins/aws/package/compile/functions/index.js

```diff
@@ -441,98 +441,125 @@ class AwsCompileFunctions {
 
-        // Read the file in chunks and add them to the hash (saves memory and performance)
-        versionCompilationPromise = BbPromise.fromCallback(cb => {
-          const readStream = fs.createReadStream(artifactFilePath);
-
-          readStream
-            .on('data', chunk => {
-              fileHash.write(chunk);
-              versionHash.write(chunk);
-            })
-            .on('close', () => {
-              cb();
-            })
-            .on('error', error => {
-              cb(error);
+        // Include function and layer configuration in version id hash (without the Code part)
+        versionCompilationPromise = this.addFileContentsToHashes(artifactFilePath, [
+          fileHash,
+          versionHash,
+        ])
+          .then(() => {
+            // Include all referenced layer code in the version id hash
+            const functionProperties = _.omit(_.get(functionResource, 'Properties', {}), 'Code');
+            const layerProperties = this.extractLayerPropertiesFromFunction(
+              functionProperties,
+              cfTemplate
+            );
+
+            const layerZipFilePaths = [];
+            layerProperties.forEach(layer => {
+              const layerZipFileName = layer.properties.Properties.Content.S3Key.split(
+                path.sep
+              ).pop();
+              const layerZipFilePath = path.join(this.packagePath, layerZipFileName);
+              layerZipFilePaths.push(layerZipFilePath);
             });
-        }).then(() => {
-          // Include function configuration in version id hash (without the Code part)
-          const properties = _.omit(_.get(functionResource, 'Properties', {}), 'Code');
-          _.forOwn(properties, value => {
-            if (_.isObject(value)) {
-              versionHash.write(JSON.stringify(value));
-            } else {
-              versionHash.write(value != null ? String(value) : '');
-            }
-          });
 
-          // Finalize hashes
-          fileHash.end();
-          versionHash.end();
+            return BbPromise.map(layerZipFilePaths, zipFileName => {
+              return this.addFileContentsToHashes(zipFileName, [versionHash]);
+            });
+          })
+          .then(() => {
+            // Include function and layer configuration in version id hash (without the Code part)
+            const functionProperties = _.omit(_.get(functionResource, 'Properties', {}), 'Code');
+            const layerProperties = this.extractLayerPropertiesFromFunction(
+              functionProperties,
+              cfTemplate
+            );
 
-          const fileDigest = fileHash.read();
-          const versionDigest = versionHash.read();
+            const mappedLayerFunctionProperties = layerProperties.map(layer => {
+              const clonedLayer = _.cloneDeep(layer);
+              clonedLayer.properties.Properties.Content.S3Key = '';
+              return clonedLayer.properties;
+            });
```

Comment:
Let's reuse already resolved `layerProperties` (nice if named `layerConfigurations`) in previous block.

Also if I see correctly this logic simply removes `S3Key` properties, that can be done simpler way via:

```javascript
for (const layerProperties of layerConfigurations) delete layerProperties.S3Key;
```

Reply:
> Let's reuse already resolved layerProperties (nice if named layerConfigurations) in previous block.

Doesn't seem to be addressed

Reply:
If I understand correctly I think this should be covered by https://github.com/serverless/serverless/pull/8066/commits/14de011386670b5ef6e78de2946b732a9197d763#diff-771a8504e96d8db95f70f2cca6e3a906R475

Reply:
It doesn't seem. Note that `layerConfigurations` are still resolved twice in code as proposed:

1. 
https://github.com/serverless/serverless/blob/bccd773804c19a6b4287502fca7fc8a9480c0597/lib/plugins/aws/package/compile/functions/index.js#L452-L455

2. 
https://github.com/serverless/serverless/blob/bccd773804c19a6b4287502fca7fc8a9480c0597/lib/plugins/aws/package/compile/functions/index.js#L476-L479

Reply:
Thanks, it is now moved out of the block and used twice.